[LimErnest] iP

LimErnest opened the pull request:
# Echo Chat Bot

The IDE I'm using is [Intellij](https://www.jetbrains.com/idea/)

> Quoting something

My **FREE** and **AMAZING** chat bot has the following features:
* add 3 types of task: to do, deadline and event
* use `list` to list all of the tasks added
* mark or unmark task that are completed
* ~~refuse to compile~~ 💀 

> [!WARNING]
> Use at your own risk. 

All you need to do is,
1. download
2. run the program
3. add tasks
4. let it manage your tasks for you  👍 

Increments:
- [x] Checkstyle
- [ ] Level-10 (coming soon)
- [ ] A-Varargs (coming soon)

Here is a sneak peak of my code:

```java

public void bye() {
        System.out.println(LINE_BREAK);
        System.out.println("Goodnight! Hope to see you again soon!");
        System.out.println("      |\\      _,,,---,,_\n" +
                           "ZZZzz /,`.-'`'    -.  ;-;;,_\n" +
                           "     |,4-  ) )-,_. ,\\ (  `'-'\n" +
                           "    '---''(_/--'  `-'\\_)");
        System.out.println(LINE_BREAK);
    }

```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

LimErnest committed:
Implementing a intital skeletal version of the chatbot

LimErnest committed:
Implemented the chatbot such that it can echo the command entered by users and exits when user types in "bye"

LimErnest committed:
Added the ability to store text entered by user and display them back when user types in "list"

LimErnest committed:
Implemented a TaskList class and a Task class and added the ability to mark tasks as done

LimErnest committed:
Added 3 different type of Tasks, ToDos, Deadlines, Events

LimErnest committed:
Use the input/output redirection technique to semi-automate the testing of Echo and also remove the tabs and spaces for the print statements

LimErnest committed:
Implemented exception handling for invalid inputs such as having no desciption, wrong commands, no deadline, no start and end time

LimErnest committed:
Added command to delete task from list

LimErnest committed:
change tag for A-TextUiTesting

LimErnest committed:
Improve all code in file and added javadocs comments

LimErnest committed:
Add feature to save tasks in the hard disk whenever task list changes and lead the data from the hard disk when the chatbot starts up

LimErnest committed:
Merge branch 'branch-Level-7'

LimErnest committed:
Add new ways to store and read dates using LocalDateTime in Event and Deadline

LimErnest committed:
Merge branch 'branch-Level-8'

LimErnest committed:
Add Parser, Storage and Ui classes for more abstraction

LimErnest committed:
Merge branch 'branch-A-MoreOOP'

LimErnest committed:
Add all classes into multiple packages

LimErnest committed:
Merge branch 'branch-A-Packages'

LimErnest committed:
Merge branch 'add-gradle-support'

LimErnest committed:
Set up gradle and ran checkstyle on code. Improved style of code according to the errors given

LimErnest committed:
Merge branch 'branch-A-Gradle'

LimErnest committed:
Add JUnit test to test methods in the following classes: Parser, Deadlines, Events, ToDos and TaskList

LimErnest committed:
Merge branch 'branch-A-JUnit'

LimErnest committed:
Package Echo Chatbot as a JAR file

LimErnest committed:
Merge branch 'branch-A-Jar'

LimErnest committed:
Add JavaDoc comments to all classes

LimErnest committed:
Improve code to comply with coding standard

LimErnest committed:
Add a new command called ""find" which allows users to find a task by searching for a keyword

LimErnest committed:
Merge branch 'branch-A-JavaDoc'

LimErnest committed:
Merge branch 'branch-A-CodingStandard'

LimErnest committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/echo/ui/Ui.java

Resolution: keep both

LimErnest committed:
Add additional messages after user terminates the chatbot for improved user experience

Include ASCII Art of a sleeping cat and changed goodbye message when user terminates the chatbot

LimErnest committed:
Add additional messages when user starts the chat bot for improved user experience

Include ASCII Art of a cat and changed welcome message when user starts the chatbot

LimErnest committed:
Add additional messages when users mark and unmark a task for improved user experience

Add words like "Meow-nificent" and "Meow-velous" to add more cat personality to the chat bot

LimErnest committed:
Add additional messages printed when user types in the wrong commands for improved user experience

Include list of valid commands and how to use them

LimErnest committed:
Add JavaDoc comment on new function showValidCommands

Add what the new function does as a JavaDoc comment and changed the ASCII Art in the greet function

LimErnest committed:
Remove full stop in comment

LimErnest committed:
Improve Javadoc comment

LimErnest committed:
Improve Javadoc comment

LimErnest committed:
Improve JavaDoc comment

LimErnest committed:
Improve Javadoc comment

LimErnest committed:
Correct code based on CheckStyle

LimErnest committed:
Merge branch 'branch-A-CheckStyle'

LimErnest committed:
Add basic JavaFX features

LimErnest committed:
Add GUI to chatbot

LimErnest committed:
Merge branch 'branch-Level-10'

chongsun2002 commented on src/main/java/echo/exception/EchoException.java:
```diff
@@ -0,0 +1,18 @@
+package echo.exception;
+
+/**
+ * Represents Exception thrown while using Echo
+ *
+ * @author Ernest Lim
+ */
+public class EchoException extends RuntimeException {
+
+    /**
+     * Constructor for EchoException object
+     *
+     * @param errorMessage message to be displayed when the error is caught
+     */
+    public EchoException(String errorMessage) {
+        super(errorMessage);
+    }
+}
```
Can this exception be more specific, and have other exceptions that inherit from it? This may help in adding new exceptions that have different behaviors.

chongsun2002 reviewed (commented):
Your code looks very good! All the components are well abstracted. I don't have much to nit on style, so I only have some suggested improvements. The code fulfils its purposes. LGTM!

C5hives commented on src/main/java/echo/main/Echo.java:
```diff
@@ -0,0 +1,233 @@
+        this.storage = new Storage(filePath);
+    }
+
+    public String greetUser() {
+        try {
+            this.storage.loadStorage(this.parser, this.taskList);
+            return ui.greet();
+        } catch (IOException e) {
+            return ui.printErrorMessage(e);
+        }
+    }
```
Is this method doing more than what its name suggests? You can consider breaking this method into two separate methods.

The method `greetUser()` currently has a side effect of loading file contents into your `taskList` and needs to handle a possible `IOException`—which is thrown from the `this.storage,loadStorage()` method and not the `ui.greet()` method.

This behavior will not be expected when other classes call `greetUser()` as the current method name makes no mention of loading tasks from a file.

You can also consider keeping the current logic as is, but renaming this method to something like `initialise()` to reduce the chances of misunderstandings when reading the code.

C5hives replied:
Is this method doing more than what its name suggests? You can consider breaking this method into two separate methods.

The method `greetUser()` currently has a side effect of loading file contents into your `taskList` and needs to handle a possible `IOException`—which is thrown from the `this.storage,loadStorage()` method and not the `ui.greet()` method.

This behavior will not be expected when other classes call `greetUser()` as the current method name makes no mention of loading tasks from a file.

You can also consider keeping the current logic as is, but renaming this method to something like `initialise()` to reduce the chances of misunderstandings when reading the code.

C5hives commented on src/main/java/echo/main/Echo.java:
```diff
@@ -0,0 +1,233 @@
+     * @param filePath String that contains the file path to store or load the list of task
+     */
+    public Echo(String filePath) {
+        this.ui = new Ui();
+        this.taskList = new TaskList();
+        this.parser = new Parser();
+        this.storage = new Storage(filePath);
```
Optional, but you can consider adhering to the advanced [coding standards](https://se-education.org/guides/conventions/java/index.html#variables) set out by this course, which is
- Avoid unnecessary use of this with fields. Use the `this` keyword only when a field is shadowed by a method or constructor parameter.

Since `ui`, `taskList`, `parser` and `storage` are not shadowed by any other parameters here, you can omit the `this` keyword (this would be applicable for every other class as well).

C5hives replied:
Optional, but you can consider adhering to the advanced [coding standards](https://se-education.org/guides/conventions/java/index.html#variables) set out by this course, which is
- Avoid unnecessary use of this with fields. Use the `this` keyword only when a field is shadowed by a method or constructor parameter.

Since `ui`, `taskList`, `parser` and `storage` are not shadowed by any other parameters here, you can omit the `this` keyword (this would be applicable for every other class as well).

C5hives commented on src/main/java/echo/task/ToDos.java:
```diff
@@ -0,0 +1,40 @@
+ *
+ * @author Ernest Lim
+ */
+public class ToDos extends Task {
```
You can consider renaming the `ToDos` class (and every other instance of it) to be singular as it represents only 1 task. It would then adhere to the [coding standard](https://se-education.org/guides/conventions/java/index.html#naming), which states that:
- Plural form should be used on names representing a collection of objects.

C5hives replied:
You can consider renaming the `ToDos` class (and every other instance of it) to be singular as it represents only 1 task. It would then adhere to the [coding standard](https://se-education.org/guides/conventions/java/index.html#naming), which states that:
- Plural form should be used on names representing a collection of objects.

C5hives commented on src/main/java/echo/main/Echo.java:
```diff
@@ -0,0 +1,233 @@
+            String command = replyArray[0];
+            String description = (replyArray.length > 1) ? replyArray[1] : "";
+
+            switch (command) {
+            case "list":
+                return handleListCommand();
+            case "mark":
+                return handleMarkCommand(description);
+            case "unmark":
+                return handleUnmarkCommand(description);
+            case "todo":
+                return handleToDoCommand(description);
+            case "deadline":
+                return handleDeadlineCommand(description);
+            case "event":
+                return handleEventCommand(description);
+            case "delete":
+                return handleDeleteCommand(description);
+            case "find":
+                return handleFindCommand(description);
+            case "bye":
+                return handleByeCommand();
+            default:
+                throw new EchoException("Sorry! I don't get what you mean. Try again with the list of commands above.");
+            }
```
Nice work on using separate methods to extract out the logic used for each command. This section of code is very easy to read! 🤩

C5hives replied:
Nice work on using separate methods to extract out the logic used for each command. This section of code is very easy to read! 🤩

C5hives commented on src/main/java/echo/tasklist/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+public class TaskList {
+
+    /** ArrayList of the tasks added to TaskList by the user */
+    private ArrayList<Task> taskList;
```
Perhaps a more intuitive variable name here? You can consider renaming `taskList` to `tasks` to indicate that it is a collection of `Task` objects.

C5hives replied:
Perhaps a more intuitive variable name here? You can consider renaming `taskList` to `tasks` to indicate that it is a collection of `Task` objects.

C5hives commented on src/main/java/echo/tasklist/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+    public ArrayList<Task> findTask(String keyword) {
+        ArrayList<Task> resultArray = new ArrayList<>();
+        for (Task task : this.taskList) {
+            String description = task.getDescription();
+            if (description.contains(keyword)) {
+                resultArray.add(task);
+            }
```
Could this logic be extracted out into the `Task` class directly? This could help with the abstraction of the `Task` class and hide its internal implementation details.

C5hives replied:
Could this logic be extracted out into the `Task` class directly? This could help with the abstraction of the `Task` class and hide its internal implementation details.

C5hives reviewed (changes requested):
The use of OOP has made the project easy to follow and read. Nice job!

C5hives reviewed (commented):
The use of OOP has made the project easy to follow and read. Nice job!
Ignore my previous PR review—I had indicated the wrong review category.

LimErnest committed:
Add assert statements and enable Java assert

LimErnest committed:
Improve code to adhere to coding standards

1. Classes have incorrect names such as ToDos, Deadlines and Events.
2. Unnecessary use of `this` with fields
3. Extract logic inside findtask(String keyword) in TaskList

1. The class only represents 1 task hence it violates the naming convention
as plural form should be used on names representing a collection of objects
2. Unnecessary noise will be created
3. Violates the Tell Dont Ask Principle

Let's,
1. rename the class to ToDo, Deadline and Event as well as every other instances of it
2. remove all unncessary use of `this` keyword
3. create another function in task called isMatch(String keyword) and
remove getDescription() method to hide its internal implementation

LimErnest committed:
Improve code to adhere to coding standards

Code to create task objects from string array is repeated

Violate Tell Dont Ask Principle as the implementation
of the logic can be hidden

Let's, change the constructor of each task class to take in
an array of string instead

Changing the parameter hides the implementation of assigning
the string array to its respective components and make the code neater

LimErnest committed:
Merge pull request #1 from LimErnest/branch-A-Assertions

Add assert statements and enable Java assert

LimErnest committed:
Merge branch 'master' of https://github.com/LimErnest/ip

* 'master' of https://github.com/LimErnest/ip:
  Add assert statements and enable Java assert

LimErnest committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/echo/tasklist/TaskList.java

LimErnest committed:
Merge pull request #2 from LimErnest/branch-A-CodeQuality

Improve coding standards

LimErnest committed:
Merge branch 'master' of https://github.com/LimErnest/ip

LimErnest committed:
Add help to list of valid command

Users have no way of knowing what are the valid commands they can use

Let's, include "help" in the switch case and implement printValidCommand() function
in Ui class. Create alert to show list of valid command when users type in "help"

Including a case for the input help and showing a pop up page of a list of valid command
allows users to know all the valid commands and its usage, improving user experience

LimErnest committed:
Merge branch 'branch-C-Help'

LimErnest committed:
Update README.md

README does not contain the information
about the chat bot.

Users would not be able to know what the
chat bot do when reading the README.

Let's update the README with images
and information about the features
of the chat bot.

Updating the README can allow users
to understand how to use the chat bot
and the features available before opening
the application.

LimErnest committed:
Update README.md

LimErnest committed:
Add method to fetch file

File path is hardcoded.

File to save and load tasklist will not be created
if file does not exist causing file not found error to
occur whenever the chat bot is started.

Let's create a method, fetchFile() to create a file
if the file is not present in the specific file path.
The method is called before loading the storage

Creating a file when the file does not exist will
not result in file not found exception and allows
user to continue saving task into tasklist.
Method is called before loading storage so that
there is for sure a file to load and save the
data into.

LimErnest committed:
Handle negative index for delete

LimErnest committed:
Save tasklist after it is changed
